src/GraphQL/Types/Scalars/ScalarGraphType.cs

```diff
@@ -41,20 +41,17 @@ public abstract class ScalarGraphType : GraphType
         /// server-side representation. Input coercion may not only return primitive values like
         /// String but rather complex ones when appropriate.
         /// <br/><br/>
-        /// This method must handle a value of <see cref="NullValue"/>.
+        /// This method must handle a value of <see cref="GraphQLNullValue"/>.
         /// </summary>
-        /// <param name="value">AST value node. Must not be <see langword="null"/>, but may be <see cref="NullValue"/>.</param>
+        /// <param name="value">AST value node. Must not be <see langword="null"/>, but may be <see cref="GraphQLNullValue"/>.</param>
         /// <returns>Internal scalar representation. Returning <see langword="null"/> is valid.</returns>
-        public virtual object? ParseLiteral(IValue value) => value switch
+        public virtual object? ParseLiteral(GraphQLValue value) => value switch
         {
-            BooleanValue b => ParseValue(b.Value.Boxed()),
-            IntValue i => ParseValue(i.Value),
-            LongValue l => ParseValue(l.Value),
-            BigIntValue bi => ParseValue(bi.Value),
-            FloatValue f => ParseValue(f.Value),
-            DecimalValue d => ParseValue(d.Value),
-            StringValue s => ParseValue(s.Value),
-            NullValue _ => ParseValue(null),
+            GraphQLBooleanValue b => ParseValue(b.ClrValue),
+            GraphQLIntValue i => ParseValue(i.ClrValue),
+            GraphQLFloatValue f => ParseValue(f.ClrValue),
+            GraphQLStringValue s => ParseValue(s.ClrValue),
```

Comment:
This code path is only used to facilitate custom scalars, and only when the custom scalar does not override `ParseLiteral` (as it should for best performance).  I would like to see this `ParseLiteral` method an abstract method, requiring it to be overridden to implement a custom scalar.  However, it is possible that people have been relying on its behavior and it is not so terrible to leave it.

I would suggest that we move all the code from the `ClrValue` properties in the parser project over here.  It seems much more appropriate to be here, when considered in relation to the other scalar graph types.

Even if it was not moved, I would certainly suggest removing the caching field from the parser project.

Reply:
I changed this code again. OK now?

Reply:
Do I need to change something else here?

Reply:
I have not reviewed but I think it’s fine. I’ll resolve conversation when reviewed.